src/java.base/share/classes/jdk/internal/classfile/impl/CodeBuilderFinalizerImpl.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.classfile.impl;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+import jdk.internal.classfile.CodeBuilder;
+import jdk.internal.classfile.Instruction;
+import jdk.internal.classfile.Label;
+import jdk.internal.classfile.TypeKind;
+import jdk.internal.classfile.instruction.BranchInstruction;
+import jdk.internal.classfile.instruction.ReturnInstruction;
+
+public final class CodeBuilderFinalizerImpl {
+
+    public static void buildWithFinalizer(CodeBuilder parent,
+                                          Consumer<CodeBuilder> tryHandler,
+                                          Consumer<CodeBuilder> finalizerHandler,
+                                          boolean compactForm,
+                                          Label... externalLabels) {
+
+        var mapping = new HashMap<Label, Label>(externalLabels.length + 1);
+        for (var extLabel : externalLabels) {
+            mapping.put(extLabel, null);
+        }
+
+        var endLabel = parent.newLabel();
+        BlockCodeBuilderImpl topBlock = new BlockCodeBuilderImpl(parent, endLabel);
+        topBlock.start();
+        var excTable = new ArrayList<Label>();
+        topBlock.transforming(
+                (cob, coe) -> {
```

Comment:
Same local/anonymous class vs lambda comment, MethodHandles does have an intrinsic (in InvokerBytecodeGenerator) for try-finally block